Added SWHID tests for cases that should fail but do not fail yet

willynilly opened the pull request:
…the current validator.

:heart: Thank you for your contribution!

### Description

> Please describe briefly your pull request.

I added some tests to address issue #121 


### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [X ] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [ X] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [X ] I've added relevant test cases.
- [ X] I've added relevant documentation.
- [ X] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/howtos/i18n/).
- [ X] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [ X] I've NOT included third-party code (copy/pasted source code or new dependencies).
    * If you have added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect on Discord](https://discord.gg/8qatqBC).

**Frontend**

- [ X] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines.
- [ X] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines.
- [ X] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines.


**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

tmorrell commented:
Thanks! In order to merge this PR we'd also need the regex to be corrected.

willynilly commented:
I just added a commit that fixes the is_swh method (including the regex) so that it complies with the SWHID standard. It now passes the tests I made in accordance with that standard.

willynilly commented:
The previous tests did not require new dependencies, but fixing the is_swh did require adding a new dependency: rfc3987.
I have not yet reached out to the architect. Can you help me with that?

tmorrell commented:
Thanks so much for these improvements! I've flagged this PR so an architect can determine whether we can include a GPL v3 licensed dependency.

max-moser commented:
thanks @willynilly for the contribution!

we've had a brief internal discussion about licenses; unfortunately the decision was to avoid GPL dependencies. :sweat: 
from our understanding, adding a GPL-licensed dependency to the inveniosoftware packages would require us to relicense our packages under the same license.
keeping the MIT license on our packages would violate the terms of the GPL license.
thus, to avoid legal risks, we cannot accept dependencies licensed under GPL or other strong copyleft licenses.

if a different, permissively licensed dependency with an implementation of the RFC exists, we could add that one instead.
alternatively, another way forward would be to implement a validator in this package (ideally without knowledge of the implementation in the rfc3987 python package), if feasible.

willynilly commented:
According to the SWID grammar (https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html#syntax), it appears that the origin need not be an absolute IRI. I changed it to be just an IRI. I also adapted some source code from a test case in the abnf-to-regexp module to recreate the functions (https://github.com/aas-core-works/abnf-to-regexp/blob/main/test_data/nested-python/rfc3987/expected.py).  The abnf-to-regexp library has an MIT license.

willynilly commented:
How do we honor the MIT license? I am new with this. https://github.com/aas-core-works/abnf-to-regexp/blob/main/LICENSE

tmorrell commented:
It looks like the regex pin would need to be sorted out https://github.com/inveniosoftware/invenio-vocabularies/blob/4f643e137964e9de2530f094672b014153f73d5b/setup.cfg#L38

willynilly commented:
can you help me understand this?

willynilly commented:
I am not importing anything from abnf-to-regexp. i used adapted some of their code

willynilly commented:
I am not importing anything from abnf-to-regexp. i used adapted some of their code. i also used their tool on a test case to generate the output that is in the same as their expected output file, which is Python code. i copied and modified this Python code. So I suppose we need to somehow copy the license to the code and include it. 

I noticed that the original checklist forbids copying code from third party sources. I tried to use existing libraries. My understanding is that MIT does allow copying and modifying source code. In this case, I used a tool to generate code, which happens to already be one of their test cases, and then I modified this code.

So I have several questions:

1) is it ok to include the code I created since it was generated by their tool and i modified it?

2) If it is not ok for the reason that you dont want copied/modified code, would it be ok if I released a python module with these functions under MIT license, and then we install/import from this third party library?

What do you think?

tmorrell commented:
Oh, sorry! I was in the review the dependencies mindset.

We'll have to get another architect review, but everything you're done seems fine.

The licensing bit is a good question. I would add Copyright (c) 2023 abnf-to-regexp AUTHORS to the top of utils.py. For now I'd also copy the MIT license text where you put the comment. I'm not sure if there is a better place to put it...so that might change.

tmorrell commented:
Oh, and also add yourself to the copyright headers for any file where you've touched more than 15 lines of code.

tmorrell reviewed (approved):
Thanks! This just needs another review and architect approval on the licensing.

slint commented on idutils/utils.py:
```diff
@@ -356,10 +358,137 @@
-swh_regexp = re.compile(
-    r"swh:1:(cnt|dir|rel|rev|snp):[0-9a-f]{40}"
-    r"(;(origin|visit|anchor|path|lines)=\S+)*$"
+"""
+MIT License
+
+Copyright (c) 2023 abnf-to-regexp AUTHORS
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+Please see the file AUTHORS for the full list of the authors of abnf-to-regexp.
+"""
+
+
+# adapted from https://github.com/aas-core-works/abnf-to-regexp/blob/main/test_data/nested-python/rfc3987/expected.py
```
I think it would be enough to instead add a small header comment here like:

```python
# The following code is adapted from abnf-to-regexp (https://github.com/aas-core-works/abnf-to-regexp)
# Copyright (c) 2023 abnf-to-regexp AUTHORS (see https://github.com/aas-core-works/abnf-to-regexp/blob/main/AUTHORS)
# License: MIT (https://opensource.org/licenses/MIT)
def _create_rfc3987_reg_exps():
    # ...
```

willynilly replied:
I did as you suggested, but did not include myself.

slint reviewed (approved):
Thanks for the contribution! Codewise/logic LGTM, just left a comment to make the license reference a bit smaller.
